### PR TITLE
refactor(docs): update 'configuring'

### PR DESCRIPTION
**Pull request by k-dovgan**
Actual purpose of this change is to differentiate the 'build step',
that is a `Step` instance, from the 'project configuration',
that is a `Configuration` instance.

### CODE REVIEW

**Comment by k-dovgan** [comment]
https://universum.readthedocs.io/en/update-configuring/

File: doc/configuring.rst [hunk 1]
```diff
@@ -18,29 +18,37 @@ to this module in `config_path` member of its input settings.
...
-Basically, `project configuration` is a single launch of some external application or script.
+Project configuration is a list of actions to be performed to test a project: e.g., prepare environment,
+build project for some specific platform or run a specific test script. These actions are mostly referred as
+"build steps". A project configuration, being a list of build steps, is sometimes referred as a build configuration.
 
-For example::
+Here's an example of such actions::
 
     $ ./build.sh -d --platform MSM8996
+    $ cp ./build/results ./tests
```
Comment (i-keliukh) [inline doc/configuring.rst:38]: Is "results" a file?

File: doc/configuring.rst [hunk 2]
```diff
@@ -18,29 +18,37 @@ to this module in `config_path` member of its input settings.
...
-to test the project: e.g., build it for some specific platform, or run some specific test script.
-Basically, `project configuration` is a single launch of some external application or script.
+Project configuration is a list of actions to be performed to test a project: e.g., prepare environment,
+build project for some specific platform or run a specific test script. These actions are mostly referred as
+"build steps". A project configuration, being a list of build steps, is sometimes referred as a build configuration.
 
-For example::
+Here's an example of such actions::
 
     $ ./build.sh -d --platform MSM8996
```
Comment (i-keliukh) [inline doc/configuring.rst:37]: Please consider changing MSM8996 to something less exotic, like linux_amd64

File: doc/configuring.rst [hunk 3]
```diff
@@ -18,29 +18,37 @@ to this module in `config_path` member of its input settings.
...
-— are three different possible project configurations.
+Each build step is defined by two main parameters:
+
+* a `name` to refer to
+* a `command` to execute
+
+Both `name` and `command`, however, can be undefined. A build step without a name will still be issued a number;
+a build step without a command will do nothing, but will still appear in log (and have a step number).
+
+Aside from these, `Universum` provides :ref:`a wide list of other helpful step parameters <keys>`.
```
Comment (i-keliukh) [inline doc/configuring.rst:50]: "wide" and "helpful" probably add to much expression to the sentence.

File: doc/configuring.rst [hunk 4]
```diff
@@ -54,33 +62,29 @@ Below is an example of the configuration file in its most basic form:
 
     configs = Variations([dict(name="Build", command=["build.sh"])])
 
-This configuration file uses a :class:`Variations` class
-from the :mod:`universum.configuration_support`
-module and defines one build configuration.
+This configuration file uses a :class:`Variations` class from the :mod:`universum.configuration_support`
```
Comment (i-keliukh) [inline doc/configuring.rst:65]: Do we plan to use ProjectConfiguration in the documentation?

Comment (o-andrieiev) [reply doc/configuring.rst:65]: Yes. I think clients should switch from Variations([dict(foo=bar)]) to Variations([ProjectConfiguration(foo=bar)]) for added type safety in their own projects. This substitution can already be done seamlessly.

**Review by o-andrieiev — APPROVED** [review]

**Comment by o-andrieiev** [comment]
Please update the commit description to reflect recent naming changes

File: setup.py [hunk 5]
```diff
@@ -38,11 +40,10 @@ def readme():
         'p4': [p4],
         'git': [git],
         'github': [github],
+        'docs': [docs],
```
Comment (i-keliukh) [inline setup.py:43]: Please add comment that 'docs' extra is needed for readthedocs

File: doc/configuring.rst [hunk 6]
```diff
@@ -18,29 +18,37 @@ to this module in `config_path` member of its input settings.
...
+Project configuration is a list of actions to be performed to test a project: e.g., prepare environment,
+build project for some specific platform or run a specific test script. These actions are mostly referred as
+"build steps". A project configuration, being a list of build steps, is sometimes referred as a build configuration.
 
-For example::
+Here's an example of such actions::
 
-    $ ./build.sh -d --platform MSM8996
+    $ ./build.sh -d --platform linux_amd64
+    $ cp ./build/results/ ./tests
```
Comment (i-keliukh) [inline doc/configuring.rst:38]: You have to add -r if you want to copy directories

File: doc/configuring.rst [hunk 7]
```diff
@@ -133,23 +137,26 @@ See the following example configuration file:
...
-List of configurations
-----------------------
+Configuration with several steps
+--------------------------------
 
-The `Universum` gets the list of project configurations from the `configs` global variable.
-In the basic form this variable contains a flat list of items, and each item represents one
-`project configuration`_.
+The `Universum` gets the list of build steps from the `configs` global variable.
+In the basic form this variable contains a flat list of items, and each item represents one`build step`_.
```
Comment (i-keliukh) [inline doc/configuring.rst:157]: Space character is missing: "one`build step`_"

File: doc/configuring.rst [hunk 8]
```diff
@@ -185,26 +189,25 @@ Each configuration description is a python dictionary with the following possibl
...
-    A project configuration can have an empty list as a command. Such configuration
-    won't do anything except informing user about missing it.
+    The list with command line for the build step launch. For every step the first list item should be
+    a console command (e.g. script name, or any other command like ``ls``), and other list items should be
+    the arguments, added to the command (e.g. ``--debug`` or ``-la``).
+    Every command line element, separated with space character, should be passed as a separate string argument.
+    Lists like ``["ls -a"]`` will not be processed correctly and therefore should be splat into ``["ls", "-a"]``.
+    Lists like ``["build.sh", "--platform A"]`` will not be processed correctly and thus should be plat into
+    ``["build.sh", "--platform", "A"]``. A build step can have an empty list as a command.
+    Such step won't do anything except informing the user about having nothing to do.
```
Comment (i-keliukh) [inline doc/configuring.rst:210]: I would rephrase it like below:
Such step won't do anything except showing up the step name in Universum execution logs.

File: doc/configuring.rst [hunk 9]
```diff
@@ -268,18 +271,20 @@ directory
...
+    Basic usage is adding ``critical=True`` to `build step`_ description.
     This parameter is used in case of a linear step execution, when the result of some step is
-    critical for the subsequent step execution. If some configuration has `critical` key set to `True`
-    and executing this step fails, no more configurations will be executed during this run.
+    critical for the subsequent step execution. If some step has `critical` key set to `True`
+    and executing this step fails, no more steps will be executed during this run.
     However, all already started :ref:`background <background_step>` steps will be finished
     regardless critical step results.
 
+.. Add proper description of nested steps `critical` functioning somewhere somehow
```
Comment (i-keliukh) [inline doc/configuring.rst:281]: Is there issue for this?

Comment (k-dovgan) [reply doc/configuring.rst:281]: No. Do we really need it? It's just a part of general documentation update, we have a lot of missing descriptions like this...

Comment (i-keliukh) [reply doc/configuring.rst:281]: Yeah, but there is a TODO in the code.
I prefer issue.

Comment (k-dovgan) [reply doc/configuring.rst:281]: #535

**Review by i-keliukh — APPROVED** [review]

File: doc/configuring.rst [hunk 10]
```diff
@@ -389,34 +396,35 @@ Below is an example of the configuration file that uses :meth:`~Configuration.du
...
-    $ ./configs.py
+    $ ./.univesrum.py
     [{'name': 'Make Special Module', 'command': 'make -C SpecialModule/', 'artifacts': 'out'},
     {'name': 'Run internal tests', 'command': 'scripts/run_tests.sh'},
     {'name': 'Run external tests', 'directory': '/home/scripts', 'command': 'run_tests.sh -d /home/Project/out/tests'}]
 
-As second and third build configurations have the same names, if log files are created,
-only two logs will be created: one for the first build step, another for both second and third,
-where the third will follow the second.
+As second and third steps have the same names, if log files are created, only two logs will be created:
```
Comment (i-keliukh) [inline doc/configuring.rst:423]: The second and third steps have different names.
Either the example or note has to be fixed.
